### ISSUE TITLE
Add support for artist artwork

### ISSUE THREAD

**Issue by tcurdt** [navidrome/navidrome#394]
At the moment my client of choice "play:sub" does show artwork for albums, but not for artists. AFAIU this should be supported if the server supports it.

![IMG_2056](https://user-images.githubusercontent.com/13697/87226241-c1a9e580-c392-11ea-83f0-ca4d825090dd.png)

The question now is - how add this to the server?
Where should the server get the artist artwork from?

I am personally not keen of using an external service for this as I want my music library to be offline capable. To me the simplest way would be an `artist.jpg` in my artist folders, just like there is the `cover.jpg` in the album folders. But others might have a different library structure?

Any thoughts how we could add this?

**Comment by deluan** [comment]
Using an external service will be implemented, but there will be configuration to disable it. Probably will integrate with Spotify or Fanart, or some similar service, which will require Navidrome user to create and configure their own API key for the service, similar to how Last.FM integration will work.

For local/offline artwork, I see 3 options:
1) Load `artist.jpg/cover.jpg/...` from the Artist folder. This assumes that the library will be organized with a structure like `.../Artist/Album/Songs.mp3`, which will not be always the case
2) Allow a configuration like `ArtistArtworkPath = "/path/to/images/{artist}.jpg"`, which Navidrome would use to find images for a given Artist. The issue with this approach is: How to handle artists with special chars in their names. Ex: AC/DC (always them :) or artists that theiir name ends with `.` (_Charlie Brown Jr._)
3) Manual upload using the UI. It is a good feature to have for occasional fixes, and will be done at some point, but would be a huge effort to add artwork for all artists of large libraries

Did you see any other options?

**Comment by tcurdt** [comment]
So the artwork lookup will basically become some kind of interface. And then there are the network and the offline based service implementations.

The service would essentially be

    func ArtistImage(name string) (URL, error)

A `LocalArtistImageResolver` could indeed use a path pattern, but you are right - special character will be a problem. On the other hand it would probably be rather safe to assume the location relative to a music track - at least on import. For every imported file N could check a defined pattern. The most likely default would be something along the lines of `../../artist.jpg`. In theory N could also just walk up the path and check for an `artist.jpg`. Feels a little simpler.

If done on import, this could build a (non-unique) index combining possible (but unlikely) different images per artist in the form of:

    encoded(artist) -> url(image)

This would even allow multiple music folders and could easily be rebuild when deleted. It does mean we would need to allow for this interface instead though:

    func ArtistImage(name string) ([]URL, error)

...which also kind of makes sense. Just not sure if that would be supported by the protocol. Otherwise just picking always the first one of the images could be an alternative/fallback, too.

This would also also allow to serve images from an upload folder. Not that I would need it :)

Does that make sense?

**Comment by tcurdt** [comment]
To bring over some information from the chat. The subsonic protocol provides the following data:
```
<subsonic-response xmlns="http://subsonic.org/restapi" status="ok" version="1.11.0">
<artistInfo2>
<biography>Black Sabbath is an English <a target='_blank' href="http://www.last.fm/tag/heavy%20metal" class="bbcode_tag" rel="tag">heavy metal</a> band that formed in 1968 in Birmingham, West Midlands, England, United Kingdom, originally comprising <a target='_blank' href="http://www.last.fm/music/Ozzy+Osbourne" class="bbcode_artist">Ozzy Osbourne</a> (vocals), <a target='_blank' href="http://www.last.fm/music/Tony+Iommi" class="bbcode_artist">Tony Iommi</a> (guitar), <a target='_blank' href="http://www.last.fm/music/Geezer+Butler" class="bbcode_artist">Geezer Butler</a> (bass), and <a target='_blank' href="http://www.last.fm/music/Bill+Ward" class="bbcode_artist">Bill Ward</a> (drums). In the early <a target='_blank' href="http://www.last.fm/tag/70s" class="bbcode_tag" rel="tag">70s</a>, they were the first to pair heavily distorted, sonically dissonant <a target='_blank' href="http://www.last.fm/tag/blues%20rock" class="bbcode_tag" rel="tag">blues rock</a> at slow speeds with lyrics about drugs, mental pain and abominations of war, thus giving birth to generations of metal bands that followed in their wake. <a target='_blank' href="http://www.last.fm/music/Black+Sabbath">Read more about Black Sabbath on Last.fm</a>.</biography>
<musicBrainzId>5182c1d9-c7d2-4dad-afa0-ccfeada921a8</musicBrainzId>
<lastFmUrl>http://www.last.fm/music/Black+Sabbath</lastFmUrl>
<smallImageUrl>http://userserve-ak.last.fm/serve/64/27904353.jpg</smallImageUrl>
<mediumImageUrl>http://userserve-ak.last.fm/serve/126/27904353.jpg</mediumImageUrl>
<largeImageUrl>http://userserve-ak.last.fm/serve/_/27904353/Black+Sabbath+sabbath+1970.jpg</largeImageUrl>
<similarArtist id="279" name="AC/DC" coverArt="ar-279" albumCount="15"/>
<similarArtist id="278" name="Accept" coverArt="ar-278" albumCount="1"/>
<similarArtist id="433" name="Kiss" coverArt="ar-433" albumCount="1"/>
<similarArtist id="372" name="Bruce Dickinson" coverArt="ar-372" albumCount="2"/>
<similarArtist id="290" name="Alice in Chains" coverArt="ar-290" albumCount="3"/>
<similarArtist id="285" name="Aerosmith" coverArt="ar-285" albumCount="2"/>
<similarArtist id="396" name="The Doors" albumCount="1"/>
<similarArtist id="397" name="Guns N' Roses" albumCount="1"/>
</artistInfo2>
</subsonic-response>
```

This means we need/should provide 3 different URLs per artist:

```
<smallImageUrl>http://userserve-ak.last.fm/serve/64/27904353.jpg</smallImageUrl>
<mediumImageUrl>http://userserve-ak.last.fm/serve/126/27904353.jpg</mediumImageUrl>
<largeImageUrl>http://userserve-ak.last.fm/serve/_/27904353/Black+Sabbath+sabbath+1970.jpg</largeImageUrl>
```

This of course will complicate things in terms of the mapping. Something along the lines of:

```
type ImageSet struct {
    Small URL
    Medium URL
    Large URL
}
func ArtistImages(name string) ([]ImageSet, error)
```

**Comment by tcurdt** [comment]
The next question is: The local resolver will only find the large image version in the filesystem. Should Navidrome generate the small and medium version on import? Should it create them on-the-fly (but cache them)?

**Comment by tcurdt** [comment]
Notes from browsing the code base:

The [ArtistInfo](https://github.com/deluan/navidrome/blob/fbd6c965b0379371e23fd1c740bab5363e5fd778/server/subsonic/responses/responses.go#L278) already has the image URLs, they just need to be set in [GetArtistInfo2](https://github.com/deluan/navidrome/blob/17df63b550404c79ef4084b5057db2364e47f9a1/server/subsonic/browsing.go#L182).

**Comment by deluan** [comment]
Yes, I've implemented a dummy version of the endpoint, just enough to make MusicStash work.

**Comment by certuna** [comment]
Artist artwork can also be embedded in the tags, like id3v2.4:
```
4.14.   Attached picture

   This frame contains a picture directly related to the audio file.
   Image format is the MIME type and subtype [MIME] for the image. In
   the event that the MIME media type name is omitted, "image/" will be
   implied. The "image/png" [PNG] or "image/jpeg" [JFIF] picture format
   should be used when interoperability is wanted. Description is a
   short description of the picture, represented as a terminated
   text string. There may be several pictures attached to one file, each
   in their individual "APIC" frame, but only one with the same content
   descriptor. There may only be one picture with the picture type
   declared as picture type $01 and $02 respectively. There is the
   possibility to put only a link to the image file by using the 'MIME
   type' "-->" and having a complete URL [URL] instead of picture data.
   The use of linked files should however be used sparingly since there
   is the risk of separation of files.

     <Header for 'Attached picture', ID: "APIC">
     Text encoding      $xx
     MIME type          <text string> $00
     Picture type       $xx
     Description        <text string according to encoding> $00 (00)
     Picture data       <binary data>


   Picture type:  $00  Other
                  $01  32x32 pixels 'file icon' (PNG only)
                  $02  Other file icon
                  $03  Cover (front)
                  $04  Cover (back)
                  $05  Leaflet page
                  $06  Media (e.g. label side of CD)
                  $07  Lead artist/lead performer/soloist
                  $08  Artist/performer
                  $09  Conductor
                  $0A  Band/Orchestra
                  $0B  Composer
                  $0C  Lyricist/text writer
                  $0D  Recording Location
                  $0E  During recording
                  $0F  During performance
                  $10  Movie/video screen capture
                  $11  A bright coloured fish
                  $12  Illustration
                  $13  Band/artist logotype
                  $14  Publisher/Studio logotype

```

**Comment by tcurdt** [comment]
Not just that. What about other formats other than mp3? Do they support that, too? And how do we get the artwork in there? Does e.g. Picard supports this?

Having to rewrite all files to change the artwork for an artist - not fantastic either.

Don't get me wrong. I see the simplicity it would provide. I am just not sure it's the most realistic or pragmatic solution.

Just playing devils advocate, I am happy to be proven wrong though :)

**Comment by certuna** [comment]
Vorbis comments use `metadata_block_picture` , and has a list of picture types similar to id3v2: https://xiph.org/flac/format.html#metadata_block_picture 
I'll have to check how it works exactly with MP4 atoms, but I'm 100% sure it can store multiple images too.

You get the images into the files, the same way as with any other metadata: a tagger.

I have to say, I'm never a huge fan of sidecar files like cover.jpg and album.jpg but I guess a lot of people are.

**Comment by tcurdt** [comment]
Well, the tagger(s) also need(s) to support it - and I cannot seem to find support in Picard for it. It supports album artwork - but I could not find a way to add artist images.

While I am not a great fan of sidecar files either, I am even less a fan of duplication - which is the other side of the same coin. 
For albums that's not so bad, for artist artwork this does feels a bit silly. Plus it raises the question - what artwork do you want to show? You have 100s of files with potentially different artist artwork and you would need to pick one image that represents the artist on the API level.

To me sidecar files are the pragmatic way, which doesn't image tags could not not be supported in the future. But TBH I just don't see a sane way how. At least not from what we need on the API level.

**Comment by certuna** [comment]
True, it does feel a bit silly to store one album image in hundreds of tracks.

Discogs btw also has artist images, and every user can generate its own API key - that's another potential source.

**Comment by tcurdt** [comment]
@certuna yes, the plan is to have multiple implementations. I am mostly interested in offline sources, but online sources are certainly of interest, too. Just a different implementation of the same interface.

**Comment by certuna** [comment]
I would probably say that deeplinking to assets in someone else's database isn't my favourite thing to do, there's no guarantee that this remote `ImageUrl` will still be there next time you try to access it. Also, will break on (temporarily) non-internet connected LANs.

Downside: local cache can get very big for large libraries.

**Comment by tcurdt** [comment]
@certuna In reality it's quite unlikely that asset URLs will change - but that is indeed an argument against caching them on import. It suggests the caching should be part of the service implementation. So a LastFM implementation could decide to not cache the URLs - or implement a cache. That is part of the question of the two approaches.

The non-internet connected LANs is exactly my focus, so I am not sure I get your point there. The URLs will then point to a local URL as in `http://navidrome.local/assets/{fingerprint}` which maps to a local file.

**Comment by deluan** [comment]
IMO the URLs (and other info) should be retrieved from the source by the service, not stored in the DB. And the service then can use the `fscache` library (used for images and transcodings) to cache the results.

For external services we need more fields, specifically MusicBrainz Artist ID, to make match more precise. This is currently not being imported by the scanner, but it will be soon.

For the LocalURL, we will need a public endpoint to expose artwork.

**Comment by tcurdt** [comment]
Using `fscache` would be nice - but I have mixed feelings. 
I guess the most simplest implementation could be along the lines of

    processChangedDir()
    .find('artist.(jpg|png)")
    .convertImageTo([small, medium, large])
    .writeToCache(encoded(artist) + size)

    ArtistImages(artist)
    .each([small, medium, large])
    .each {
       return foundInCache(encoded(artist) + size)
       ? encoded(arist) + size
       : noURL
    }

    http://navi/art/{key}
    response = cache.Get(key)

Things get a bit more complicated though when there are multiple images per artist - if that is a requirement.
The cache entry would need to become a mutable collection. That would make this approach not such a great idea IMO.

I am by no means eager to store this in the DB, but maybe you could outline the approach you have in mind a bit more, @deluan ? Also at what stage do you see the conversion into the various sizes happening?

**Comment by deluan** [comment]
I think the offline/disconnected service could be implemented fully at query time (i.e. not as part of the scan process). Whenever the `getArtistInfo(2)` endpoint is called, it tries to read the image from the artist folder. The challenge is determining what is the artist folder in cases where your albums are spread across different folders (my case, actually). 

I'll focus on the external (probably Spotify) service this week, so I'll have a clearer picture of the overall architecture.

> Also at what stage do you see the conversion into the various sizes happening?

I think we need only one image (large) in the folder: the resizing will be handled by the endpoint that provides the image, we just need to pass the size as a query param (ex: `?size=300`). This is how the current Subsonic API `getCoverArt` works. And it also caches the resized image.

**Comment by tcurdt** [comment]
> I think the offline/disconnected service could be fully implemented at query time (i.e. not as part of the scan process). Whenever the `getArtistInfo(2)` endpoint is called, it tries to read the image from the artist folder.

The one benefit I see here is that it would not need some update schedule to fix the covers.

> The challenge is determining what is the artist folder in cases where your albums are spread across different folders (my case, actually).

I think that approach is a bit tricky.
The very least you would then need a mapping to the various locations where to check. A pattern would not work in this case. And extracting the locations from all the individual tracks does not sound like a great idea.

I don't think we can get away without storying at least something to make this a little easier. Be it just a list of directories inside the db.  But I am looking forward in what you can come up with.

> I'll focus on the external (probably Spotify) service this week, so I'll have a clearer picture of the overall architecture.

Well, feel free to reach out on discord if you want to chat.

Another thing that I was thinking about is a combination of the online/offline world. 
What about caching the responses of the external Service? Maybe even caching the Images?
That would gets us already quite close the offline world.

I would still prefer to have the artwork being a permanent part of my library though.
Especially as not all artist will be have artwork on *whatever external service*.

> > Also at what stage do you see the conversion into the various sizes happening?
> 
> I think we need only one image (large) in the folder: the resizing will be handled by the endpoint that provides the image, we just need to pass the size as a query param (ex: `?size=300`). This is how the current Subsonic API `getCoverArt` works. And it also caches the resized image.

I think that's fair. That would certainly make some parts easier.

**Comment by deluan** [comment]
Hey @tcurdt I'm thinking about creating a set of interfaces that will allow easy implementation of future integrations (including local artist images):
```
package agents

import (
	"context"
)

type Artist struct {
	Name      string
	MBID      string
	Biography string
	URL       string
}

type Song struct {
	Name string
	MBID string
}

type ArtistInfoRetriever interface {
	GetArtistInfo(ctx context.Context, artistName, artistMBID string) (*Artist, error)
}

type SimilarArtistsRetriever interface {
	GetSimilarArtists(ctx context.Context, artistName, artistMBID string, limit int) ([]Artist, error)
}

type ArtistImagesRetriever interface {
	GetArtistImages(ctx context.Context, artistName, artistMBID string) ([]string, error)
}

type ArtistTopSongsRetriever interface {
	GetArtistTopSongs(ctx context.Context, artistName, artistMBID string, limit int) ([]Song, error)
}
```

Then an "orchestrator", configured with a prioritized list of agents for each one of those areas (info, similar, images, topSongs), would call each implementation until it gets all information or exhaust the possibilities, then combine all results into an `ArtistInfo` struct:

```
type ArtistInfo struct {
	ID             string
	Name           string
	MBID           string
	Biography      string
	SmallImageUrl  string
	MediumImageUrl string
	LargeImageUrl  string
	ArtistUrl      string
	SimilarArtists Artists
}

This segregation of interfaces would allow to implement only the parts provided by a specific API. Ex: Only images from local, Spotify could provide images and topSongs, Last.FM could provide everything except images, etc.... 

What do you think?

**Comment by tcurdt** [comment]
Sorry for the late reply. I am excited about the progress :)

Thinking about a local implementations I don't see any particular problems. Two things:

1. I am wondering why you chose `string` over `URL` though. Any particular reason?

2. The configuration/instantiation of the agents would be an interesting detail to know as well.

**Comment by deluan** [comment]
1. No particular reason, but it is a kind of opaque data we don't have any need to parse, so I don't see the point in converting it to string. But could work either way
2. The idea is to have it defined in the configuration, and the "orchestrator" would instantiate each agent and use them as defined in the config and based on which interface each agent implements

**Comment by tcurdt** [comment]
I know you said you are a bit busy atm, but I just wanted to check about the current state of things.

**Comment by deluan** [comment]
Yes, didn't forget about this. I'm back at dedicating more time to Navidrome and this is in my top 3 action itens

**Comment by deluan** [comment]
@tcurdt Take a look at the description of the PR #766 for details on how agents will be implemented

**Comment by tcurdt** [comment]
Just a first (very!) quick look.

I think in the constructor we should pass an array of paths/patterns.
That should then also be future proof for multi library support.

It feels a bit redundant but how about also passing the base URL in via constructor (for now)?

Do you think the agent should register the endpoint?

**Comment by certuna** [comment]
How do embedded Artist images in the file tags fit into this? The tag formats (id3, mp4, Vorbis) all support it, will Navidrome?

One thing that annoys me with Plex for example is that it pulls in Mood from last.fm/Musicbrainz, but doesn't read the Mood fields from the tags, breaking the "embedded metadata always trumps online metadata" paradigm.

**Comment by deluan** [comment]
@certuna I agree with you and tags should trump online metadata and any sidecar files. But I could not find a way to add artist pictures to files, I tried with Picard and iTunes. And as I don't have any files with these pictures, I cannot even check if `ffmpeg` and/or `taglib` support extracting them.

Can you please share with me how are you adding these tags? Or at least send me 1 file of each tag type (id3, vorbis, mp4) so I can test it here?

**Comment by deluan** [comment]
@tcurdt , paths can be get from the DB, using `DataStore.MediaFolder` that can be injected in your agent if we use wire to create the constructor. My concern is the base URL, as we don't have that anywhere in the server. We may need to introduce it as part of the `BaseURL` config option

**Comment by certuna** [comment]
[Samples.zip](https://github.com/navidrome/navidrome/files/6142378/Samples.zip)
Here's three samples of files with both embedded cover art *and* artist images.

id3v2.4 has these picture types defined, my proposal would be to map 7, 8, A, E, F and 13 (in that order of preference) to Navidrome's artist picture object:
```
$00  Other
$01  32x32 pixels 'file icon' (PNG only)
$02  Other file icon
$03  Cover (front)
$04  Cover (back)
$05  Leaflet page
$06  Media (e.g. label side of CD)
$07  Lead artist/lead performer/soloist
$08  Artist/performer
$09  Conductor
$0A  Band/Orchestra
$0B  Composer
$0C  Lyricist/text writer
$0D  Recording Location
$0E  During recording
$0F  During performance
$10  Movie/video screen capture
$11  A bright coloured fish
$12  Illustration
$13  Band/artist logotype
$14  Publisher/Studio logotype
```

**Comment by tbleiker** [comment]
What is the current state on providing the artist artwork from local storage (e.g. artist.jpg in artist folder)? I have some artists which are not on Spotify and I would like to get an image for them as well 😄

**Comment by tcurdt** [comment]
IIUC @deluan has prepared the interface but implement only a remote lookup. I didn't get around to look into making the local lookup happen yet. But it should be possible to implement now - that's at least how I remember it, @tbleiker

**Comment by tcurdt** [comment]
I just looked at https://github.com/navidrome/navidrome/blob/master/core/agents/placeholders.go

Seems we still need an endpoint to expose images from a local cache.
Should that become part of the agent? @deluan 
Does that maybe need more lifecycle methods than just `init`?

**Comment by poland153** [comment]
> @certuna I agree with you and tags should trump online metadata and any sidecar files. But I could not find a way to add artist pictures to files, I tried with Picard and iTunes. And as I don't have any files with these pictures, I cannot even check if `ffmpeg` and/or `taglib` support extracting them.
> 
> Can you please share with me how are you adding these tags? Or at least send me 1 file of each tag type (id3, vorbis, mp4) so I can test it here?

Just my 2 cents: I am able to embed artist artwork into a file using foobar2000. That way the album front cover is embedded as 1/2 and the artist is embedded as 2/2. 
(Front Cover)
![image](https://user-images.githubusercontent.com/10625890/162592007-e5a55482-0b9e-440e-913e-6466d4058325.png)
(Artist)
![image](https://user-images.githubusercontent.com/10625890/162592013-c23ce847-b74c-4087-a7ef-0b9124f7d753.png)

mp3tag works okay-ish but in my experience it was a bigger struggle doing updates across lots of files compared to foobar2000.  For example, I can right click on the main Lady Gaga folder, click Properties, and under Artwork tab itll show everything that it can set. Obviously the front covers will be different, but its still able to explain that they have front covers. 
![image](https://user-images.githubusercontent.com/10625890/162591745-bdb3fd07-dfc9-4844-a1d8-84baac332288.png)

I wasnt able to upload the flac file since github 1. doesnt accept flac 2. might be too big of a file. Instead i uploaded it to Send.vis.ee since Firefox Send is dead (rip). Unfortunetly the link does die 20 downloads or 7 days since upload. So if its missed I can reupload. https://send.vis.ee/download/3a4a06aa66ecc365/#SpWfweiyVLUqCLQ1gAJpag

**Comment by airikr-e** [comment]
> Using an external service will be implemented at some point, but no worries: there will be configuration to disable it. Probably will integrate with Spotify or Fanart, or some similar service, which will require Navidrome users to create and configure their own API key for the service, similar to how Last.FM integration will work.
> 
> For local/offline artwork, I see 3 options:
> 
>     1. Load `artist.jpg/cover.jpg/...` from the Artist folder. This assumes that the library will be organized with a structure like `.../Artist/Album/Songs.mp3`, which will not be always the case
> 
>     2. Allow a configuration like `ArtistArtworkPath = "/path/to/images/{artist}.jpg"`, which Navidrome would use to find images for a given Artist. The issue with this approach is: How to handle artists with special chars in their names. Ex: AC/DC (always them :) or artists that theiir name ends with `.` (_Charlie Brown Jr._)
> 
>     3. Manual upload using the UI. It is a good feature to have for occasional fixes, and will be done at some point, but would be a huge effort to add artwork for all artists in large libraries
> 
> 
> Did you see any other options?

Both the second and the third option seemed like good options 🙂 Now almost 2 years after, howis the status of this?

**Comment by tcurdt** [comment]
To implement the lookup should be easy. And option 2 seems to most straight forward.
We could even chain the lookups and just have a cache folder.

  URL -> Cache -> Spotify/Fanart

So building the offline image gallery would be just listening to a song per artist.

I think the biggest difference here is that Navidrome would need to serve the images instead of handing out URLs to remote services. I think that's the major reason the issue got stale.

Maybe @deluan could weigh in here.

**Comment by tcurdt** [comment]
Since this issue has gone stale...

How about doing the following:

1. Have an endpoint `http://navidrome.local/assets/{fingerprint}` that serves images from a cache (without auth)
2. Calculate the `fingerprint` from the artist name and a random seed (ideally every artist should have its own seed)
3. The current lookup will store the resolved result inside the cache

Some other nice to have would then be some way to tell navidrome "go fetch the information for all artists in the db".

Another totally different option could be to have a file based lookup that comes with its own http endpoint to serve the files.
Not sure how that would fit into the navidrome software architecture though.

@deluan please advice how to move forward with this.

**Comment by deluan** [comment]
Hey @tcurdt sorry for letting this become stale. After re-reading the thread and the code, I think the challenge is still the same:

> My concern is the base URL, as we don't have that anywhere in the server. We may need to introduce it as part of the BaseURL config option (set it to https://my.server.com/music instead of /music) or create a new config option... We can potentially try to extract it from the client's http request, though it's not reliable

Any good ideas on how to solve this issue?

**Comment by tcurdt** [comment]
I think it should be as little magic as possible.

For this feature to work we need to know the full URL where the server is accessible.
If there is no full BaseURL the feature cannot be used.
As simple as that.

So the default config could just work the way it is right now.
But if we provide full BaseURL via config, the images can be cached and delivered through Navidrome.

That said....

Having a BaseURL be a required option in the config would also feel like a common thing to expect.
Just would be a breaking change.

**Comment by StoneEngineer** [comment]
Yesterday I installed Navidrome on TrueNAS VM with library on ZFS dataset accessed from VM via NFS share.
Android player accessing Navidrome from anywhere privately through VPN + TLS via reverse proxy...
Works flawlessly!

I would welcome minimalistic approach at the moment... artist image along the audio file on the server.
smallImageUrl, mediumImageUrl, largeImageUrl... same file or generated on-the-fly and chached (switchable option).
I like the proposal by tcurdt above... https://github.com/navidrome/navidrome/issues/394#issuecomment-679197711
IMO... no database, no multiple images per artist, caching (time > storage), base url or /artist/artwork

Thaks again for Navidrome!

**Comment by emes81** [comment]
I've just read through this thread and would have one consideration to add: for option 1 in deluan's post at the top of this thread, how would Navidrome handle artist images for artists who are present only on compilations? For that reason alone, option 2 seems to me much more robust

**Comment by tcurdt** [comment]
Can you please quote or link the "option 2" you are referring to? Makes it a little hard to follow otherwise.

Compilations or not - you could still have an artist folder for those artists that appear in the compilation.

**Comment by tcurdt** [comment]
Fair point. It would also probably align itself more to the idea of being a cache for images.
And if there ever might come support for multiple music folders - the cache could even be shared.

**Comment by tcurdt** [comment]
I haven't look into it in a while but it seems like the biggest obstacle is to expose that cache on the navidrome URL space.

It is also a bit of a privacy issue to expose that cache. But thinking of it as a cache - maybe we don't even need authentication then?

If there is an artist in the db we would store a copy, if there is no artist in the db we would just proxy the image.
So there would no longer be a way to tell if you have that artist in your db or not.
That would make things considerably less complex.

**Comment by deluan** [comment]
There's a draft PR for this in https://github.com/navidrome/navidrome/pull/1921. I'll work with @kgarner7 to complete this for next release.

> If there is an artist in the db we would store a copy, if there is no artist in the db we would just proxy the image.
> So there would no longer be a way to tell if you have that artist in your db or not.

That's not a bad idea! :)

**Comment by altosys** [comment]
Just chiming in here to say that option 1 above aligns well with how the images are located when managing a music library with [beets](https://beets.io). Perhaps there should be a toggle option to use files found in a structured library like that, or just tell Navidrome to dump all images in a single cache folder if the library is unstructured.

**Comment by deluan** [comment]
Hey @altosys , I'm implementing this right now. I use beets but I'm not familiar with this naming schema you mentioned. Is it a specific plugin that provides artist images? 

The current implementation I'm working on is to look for an image file named `artist.*` (literally, not the artist name) in the artist's albums common parent folder or in any of the artist's albums folders:
```
/library/U2/artist.png                                <- THIS HAS PRIORITY
/library/U2/Zooropa/01 Zooropa.flac
/library/U2/Zooropa/artist.jpg                         <- THIS IS THE SECOND OPTION
/library/U2/Boy/01 I Will Follow.mp3
/library/U2/Boy/02 Twilight.mp3
```

**Comment by altosys** [comment]
@deluan I see that you already merged and closed this, but just for reference beets can use [this](https://github.com/dkanada/beets-fetchartist) plugin for fetching the artists. It doesn't seem to have a default filename but I think jellyfin might have used `poster.jpg`.

I think using `artist.jpg` makes the most sense anyway so I'm happy to see this merged. When can we expect this to be released?

**Comment by deluan** [comment]
@altosys I thought you were using some other plugin, as this one should not work anymore: Last.fm does not provide artist images anymore: https://stackoverflow.com/a/59529682

**Comment by altosys** [comment]
> @altosys I thought you were using some other plugin, as this one should not work anymore: Last.fm does not provide artist images anymore: https://stackoverflow.com/a/59529682

It works because it doesn't use the API, as mentioned at the top of the README:

```
A plugin that fetches artist covers from last.fm and places them in the artist directories. They removed this functionality from their API a while ago so this library is parsing the HTML and manually grabbing the images off their site.
```

**Comment by Igorgro** [comment]
Does the fixing this issue also means that this https://github.com/navidrome/navidrome/issues/1869 will be also closed?

**Comment by deluan** [comment]
@Igorgro I haven't look into #1869 yet, so I don't expect it to be already fixed.

**Comment by altosys** [comment]
> Hey @altosys , I'm implementing this right now. I use beets but I'm not familiar with this naming schema you mentioned. Is it a specific plugin that provides artist images?
> 
> The current implementation I'm working on is to look for an image file named `artist.*` (literally, not the artist name) in the artist's albums common parent folder or in any of the artist's albums folders:
> 
> ```
> /library/U2/artist.png                                <- THIS HAS PRIORITY
> /library/U2/Zooropa/01 Zooropa.flac
> /library/U2/Zooropa/artist.jpg                        <- THIS IS THE SECOND OPTION
> /library/U2/Boy/01 I Will Follow.mp3
> /library/U2/Boy/02 Twilight.mp3
> ```

Coming back to this naming, I realized that I got the name `poster.jpg` from the Jellyfin music library artist [filename requirements](https://jellyfin.org/docs/general/server/media/music/#primary) where `artist` is not one of the supported options.

This is a shame since now I'm using beets to fetch `artist.jpg`, which works for navidrome, but not for Jellyfin. Would it be feasible to consider supporting the same list of filenames as Jellyfin does in addition to `artist`?

**Comment by deluan** [comment]
> Would it be feasible to consider supporting the same list of filenames as Jellyfin does in addition to `artist`?

I may introduce something like `ArtistArtPriority` (similar to `CoverArtPriority`), where you can specify the names of files to look for, as well as the order of art lookups (ex: local files first, then external sources).

**Comment by altosys** [comment]
> I may introduce something like `ArtistArtPriority` (similar to `CoverArtPriority`), where you can specify the names of files to look for, as well as the order of art lookups (ex: local files first, then external sources).

That sounds excellent

**Comment by bbsixzz** [comment]
1. The artwork doesn't work for artist with a space in their name, e.g. Moby works but Aphex Twin doesn't.

2. When I remove or replace an artist.* file the old one get's stuck and I can't get it to update unless I relog.

**Comment by deluan** [comment]
> 1. The artwork doesn't work for artists with a space in their name, e.g. Moby works but Aphex Twin doesn't.
>    [EDIT] It's actually artist folders containing folders with `\` in their name.

Can you please create a new issue and add the full path for the `artist.jpg` file? This one is already closed and it is a feature request, not a bug report. Thanks!

> 2. When I remove or replace an artist.* file the old one get's stuck and I can't get it to update unless I relog.

Yeah, that's because the browser caches the image. There's nothing much I can do for now...

**Comment by ghenry22** [comment]
> > Would it be feasible to consider supporting the same list of filenames as Jellyfin does in addition to `artist`?
> 
> I may introduce something like `ArtistArtPriority` (similar to `CoverArtPriority`), where you can specify the names of files to look for, as well as the order of art lookups (ex: local files first, then external sources).

This would be an excellent addition to make the artwork source predictable.

**Comment by deluan** [comment]
> This would be an excellent addition to make the artwork source predictable.

Well, it **_is_** predictable now: https://www.navidrome.org/docs/usage/artwork/#artists (unless you add `artist.jpg` in more then one album folder)

What this option would bring is more flexibility.
